# Exercise 2.A Fundamentals of pandas

import numpy and pandas

In [1]:
import numpy as np 
import pandas as pd

## Series

Create four new variables named, aa, ab, ac, and ad to hold plain objects

In [2]:
# assign aa the list 1,2,3,4
aa = [1,2,3,4]

In [3]:
# assign ab the tuple 5,6,7,8
ab = (5,6,7,8)

In [4]:
# assign ac the set 9,10,11,12
ac = {9,10,11,12}

In [5]:
# assign ad a dictionary with key-value pairs:
#'first': 13, 'second': 14, 'third': 15, 'fourth': 16
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}

Create four new variable named ser1, ser2, ser3, and ser4.

In [6]:
# Ser1 should use aa
ser1 = pd.Series(aa)
ser1

0    1
1    2
2    3
3    4
dtype: int64

In [7]:
# Ser2 should us ab
ser2 = pd.Series(ab)
ser2

0    5
1    6
2    7
3    8
dtype: int64

In [8]:
# Ser3 should use ac
ser3 = pd.Series(list(ac))
ser3


0     9
1    10
2    11
3    12
dtype: int64

In [9]:
# Ser4 should use ad
ser4 = pd.Series(ad)
ser4

first     13
second    14
third     15
fourth    16
dtype: int64

Create a variable ser5 that assigns the literal values 17, 18, 19, 20

In [10]:
ser5 = pd.Series(17,18,19,20)

# It gives a TypeError, The error message is Series.init()

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

Create a new variable df1 and give it the value of a DF using ser1

In [11]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


Add two new columns to df1 via insert() using ser2 and ser3

In [12]:
df1.insert(1, "ser2", ser2)
df1.insert(2, "ser3", ser3)
df1

,0,ser2,ser3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


Add a third column to df1 using ser4

In [13]:
df1.insert(3, "ser4", ser4)
df1

# After calling the DataFrame, the values in column: ser4, all shows NaN. This happened because ser4 has a dictionary collection in it. It has string values that don't match the other values in the DF

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


Add a new row 4 using the loc attribute with the values 5, 9, 12, 0

In [14]:
df1.loc[4] = [5,9,12,0]
df1

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

Create a new variable df2. Use the read_csv function to read the contents of sakila into df2

In [17]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


Display the top 5 records

In [18]:
df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


Display the last 10 records

In [19]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


Show a summary information about df2

In [20]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


I can see in the info results that all the values are string type. Additionally they all have the same amount of non-null count.

Show column statistics on df2

In [21]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


I've noticed that it now has unique, top and freq. It shows more depth of information for the values in the DF. Like how of the 262 titles, only 46 of them are unique.

Create a new variable df3, that reads the content of northwind file

In [23]:
df3 = pd.read_csv('northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


Show column statistics

In [24]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


A useful insight I see from the results are the mean values especially for UnitPrice. The summarization here is different from the previous, because it is summarizing values that are integer type. It shows min/max and percentiles.

## Coding a DataFrame

Create a new variable test_data and assign it a dictionary

In [25]:
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

Create a DF named df4 using test_data.

In [26]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


Export df4 to a CSV file named test_data.csv

In [27]:
df4.to_csv('test_data.csv', index = False)